plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -2829,98 +2856,126 @@ void producer_plugin_impl::start_write_window() {
    _ro_read_window_timer.expires_from_now( expire_time );
-   _ro_read_window_timer.async_wait( app().executor().wrap(  // stay on app thread
+   // Needs to be on read_only because that is what is being processed until switch_to_write_window().
+   _ro_read_window_timer.async_wait( app().executor().wrap(
       priority::high,
-      exec_queue::read_only_trx_safe,
+      exec_queue::read_only,
       [weak_this = weak_from_this()]( const boost::system::error_code& ec ) {
          auto self = weak_this.lock();
          if( self && ec != boost::asio::error::operation_aborted ) {
             // use future to make sure all read-only tasks finished before switching to write window
-            for ( auto& task: self->_ro_trx_exec_tasks_fut ) {
+            for ( auto& task: self->_ro_exec_tasks_fut ) {
                task.get();
             }
-            self->_ro_trx_exec_tasks_fut.clear();
+            self->_ro_exec_tasks_fut.clear();
+            // will be executed from the main app thread because all read-only threads are idle now
             self->switch_to_write_window();
           } else if ( self ) {
-             self->_ro_trx_exec_tasks_fut.clear();
+             self->_ro_exec_tasks_fut.clear();
           }
        }));
 }
 
-// Called from a read only trx thread. Run in parallel with app and other read only trx threads
-bool producer_plugin_impl::read_only_trx_execution_task() {
-   auto start = fc::time_point::now();
-   auto read_window_deadline = start + _ro_read_window_effective_time_us;
+// Called from a read only thread. Run in parallel with app and other read only threads
+bool producer_plugin_impl::read_only_execution_task() {
    // We have 4 ways to break out the while loop:
    // 1. pass read window deadline
-   // 2. Net_plugin receives a block
-   // 3. No more transactions in the read-only trx queue
-   // 4. A transaction execution is exhaused
-   while ( fc::time_point::now() < read_window_deadline && !_received_block ) {
-      std::unique_lock<std::mutex> lck( _ro_trx_queue.mtx );
-      if ( _ro_trx_queue.queue.empty() ) {
-         break;
-      }
-      auto trx = _ro_trx_queue.queue.front();
-      _ro_trx_queue.queue.pop_front();
-      lck.unlock();
-      
-      auto retry = process_read_only_transaction( trx.trx, trx.next, start );
-      if ( retry ) {
-         lck.lock();
-         _ro_trx_queue.queue.push_front(trx);
-         // Do not schedule new execution
-         break;
+   // 2. net_plugin receives a block
+   // 3. No more transactions in the exhausted read-only queue and no read-only tasks to execute
+   // 4. A transaction execution is exhausted (means end of read window)
+   bool exhausted_trx_queue_empty = false;
+   while ( fc::time_point::now() < _ro_window_deadline && !_received_block && !_ro_exiting_read_window ) {
+      if (!exhausted_trx_queue_empty) {
+         ro_trx_t t;
+         exhausted_trx_queue_empty = !_ro_exhausted_trx_queue.pop_front(t);
```

Comment:
It seems that we aren't pushing ro transactions to the `_ro_exhausted_trx_queue` queue anymore (so we won't process them in parallel).

Reply:
They are pushed into the read_only task queue and run in parallel out of there along with other read_only tasks.

Reply:
so why do we even need `_ro_exhausted_trx_queue`?

Reply:
I considered just removing and re-posting to the read-only queue. 

However, we currently use the unapplied transaction queue for non-read-only trxs instead of re-posting. That keeps us from processing them more than once per block. My understanding is the original design it was decided to keep them out of the unapplied trx queue and have their own queue. So to mirror that behavior I kept the exhausted read-only queue.

Reply:
Yeah, it should be once per read-window. So the queue keeps us from doing that. Without the queue (or some mechanism) we would potentially try it again in the same read-window.

Reply:
Not really. Once in the `_ro_exhausted_trx_queue`, they could be tried multiple times within a read window, right?

Reply:
No, once we hit exhausted on a read-only we exit the read-window.